ntt/kernels.cu

```diff
@@ -41,61 +32,101 @@ void swap(T& u1, T& u2)
     u2 = temp;
 }
 
+template<typename T>
+__device__ __forceinline__
+T bit_rev(T i, unsigned int nbits)
+{
+    if (sizeof(i) == 4 || nbits <= 32)
+        return __brev(i) >> (8*sizeof(unsigned int) - nbits);
+    else
+        return __brevll(i) >> (8*sizeof(unsigned long long) - nbits);
+}
+
 // Permutes the data in an array such that data[i] = data[bit_reverse(i)]
 // and data[bit_reverse(i)] = data[i]
 __launch_bounds__(1024) __global__
 void bit_rev_permutation(fr_t* d_out, const fr_t *d_in, uint32_t lg_domain_size)
 {
-    index_t i = threadIdx.x + blockDim.x * (index_t)blockIdx.x;
-    index_t r = bit_rev(i, lg_domain_size);
-
-    if (i < r || (d_out != d_in && i == r)) {
-        fr_t t0 = d_in[i];
-        fr_t t1 = d_in[r];
-        d_out[r] = t0;
-        d_out[i] = t1;
+    if (gridDim.x == 1 && blockDim.x == (1 << lg_domain_size)) {
+        uint32_t idx = threadIdx.x;
+        uint32_t rev = bit_rev(idx, lg_domain_size);
+
+        fr_t t = d_in[idx];
+        __syncthreads();
```

Comment:
Fixed. Thanks! For reference, **formally** speaking the correct condition is "overlapped," not just "fully overlapped." I mean `d_out != d_in` is not an actual guarantee that bad things can't happen. Because it doesn't cover the "partially overlapped" case ;-) But for all practical reasons it's good enough. Well, unless it's guaranteed that it confuses the cache out of efficiency... But it's nvidia's problem, not ours, right?

Reply:
> But it's nvidia's problem, not ours, right?

I guess, but those can quickly become a problem for us too :) Here the domain size is so small that it wouldn't make any meaningful difference, so I think we are fine to go with readability regardless :)